### PR TITLE
SER-3032 | Implement other FileBackend operations for GCS

### PR DESCRIPTION
**Pull request by slayful**

### CODE REVIEW

File: includes/filerepo/backend/GcsFileBackend.php [hunk 1]
```diff
@@ -1,43 +1,82 @@
...
-	 * $params include:
-	 *     src    : source storage path
-	 *     latest : use the latest available data
-	 *
-	 * @param $params Array
-	 * @return TempFSFile|null Returns null on failure
+	 * Copied from SwiftFileBackend for compatibility. Might have no sense.
+	 * @see FileBackendStore::isValidContainerName()
+	 */
+	protected static function isValidContainerName( $container ) {
```
Comment (pwojcik86) [inline includes/filerepo/backend/GcsFileBackend.php]: Not neccessary

File: includes/filerepo/backend/GcsFileBackend.php [hunk 2]
```diff
@@ -1,43 +1,82 @@
...
+	 * Copied from SwiftFileBackend for compatibility. Might have no sense.
+	 * @see FileBackendStore::resolveContainerPath()
+	 * @param $container
+	 * @param $relStoragePath
+	 * @return string|null
+	 */
+	protected function resolveContainerPath( $container, $relStoragePath ) {
+		if ( !mb_check_encoding( $relStoragePath, 'UTF-8' ) ) { // mb_string required by CF
+			return null; // not UTF-8, makes it hard to use CF and the swift HTTP API
+		} elseif ( strlen( urlencode( $relStoragePath ) ) > 1024 ) {
```
Comment (pwojcik86) [inline includes/filerepo/backend/GcsFileBackend.php]: 👍

File: includes/filerepo/backend/GcsFileBackend.php [hunk 3]
```diff
@@ -1,43 +1,82 @@
...
+			return null; // too long for Swift
```
Comment (pwojcik86) [inline includes/filerepo/backend/GcsFileBackend.php]: GCS

File: includes/filerepo/backend/GcsFileBackend.php [hunk 4]
```diff
@@ -1,43 +1,82 @@
...
+
+	/**
+	 * Copied from SwiftFileBackend for compatibility. Might have no sense.
+	 * @see FileBackendStore::resolveContainerPath()
+	 * @param $container
+	 * @param $relStoragePath
+	 * @return string|null
+	 */
+	protected function resolveContainerPath( $container, $relStoragePath ) {
+		if ( !mb_check_encoding( $relStoragePath, 'UTF-8' ) ) { // mb_string required by CF
```
Comment (pwojcik86) [inline includes/filerepo/backend/GcsFileBackend.php]: We do not use CloudFiles here

File: includes/filerepo/backend/GcsFileBackend.php [hunk 5]
```diff
@@ -51,59 +90,208 @@ public function getLocalCopy( array $params ) {
...
+			'params' => json_encode( $params ),
+			'sha1' => $sha1,
+		] );
 
 		try {
 			$this->bucket()->upload( fopen( $params['src'], 'r' ), [
-				'name' => $this->objectName( $params['dst'] )
+				'name' => $this->objectName( $params['dst'] ),
+				'metadata' => $this->getMetadata( $sha1 ),
+				'metadata.zorf' => 'morf',
```
Comment (pwojcik86) [inline includes/filerepo/backend/GcsFileBackend.php]: Not needed

File: includes/filerepo/backend/GcsFileBackend.php [hunk 6]
```diff
@@ -112,21 +300,35 @@ protected function doDeleteInternal( array $params ) {
...
+	private function getMetadata( $sha1 ) {
+		return [
+			// user metadata are nested like so: metadata.metadata
+			'metadata' => [
+				'sha1' => $sha1,
+			]
+		];
+	}
+
+	private function sha1ToHash( $sha1 ) {
```
Comment (pwojcik86) [inline includes/filerepo/backend/GcsFileBackend.php]: I don't see where this is used

**Review by pwojcik86 — CHANGES_REQUESTED** [review]
Could you write some tests as well?

File: includes/filerepo/backend/GcsFileBackend.php (hunk 6 above)
Comment (shaedrich) [reply includes/filerepo/backend/GcsFileBackend.php]: https://github.com/Wikia/app/pull/16732/files#diff-dee15708210663e64e57b83e68d93219R312

File: includes/filerepo/backend/GcsFileBackend.php (hunk 2 above)
Comment (mszabo-wikia) [reply includes/filerepo/backend/GcsFileBackend.php]: as it's utf-8, should we use `mb_strlen` (character count)?

File: includes/filerepo/backend/GcsFileBackend.php [hunk 7]
```diff
@@ -1,43 +1,89 @@
...
+	 * @param $container
+	 * @param $relStoragePath
+	 * @return string|null
+	 */
+	protected function resolveContainerPath( $container, $relStoragePath ) {
+		if ( strlen( $relStoragePath ) == 0 && strlen( $container ) == 0 ) {
+			return null; // path and container cannot be empty
+		} elseif ( !mb_check_encoding( $relStoragePath, 'UTF-8' ) ) {
+			return null; // not UTF-8, not supported by GCS
+		} elseif ( strlen( urlencode( $container ) ) + strlen( urlencode( $relStoragePath ) ) >
```
Comment (pwojcik86) [inline includes/filerepo/backend/GcsFileBackend.php]: `$this->gcsPaths->objectName` should be used here

**Review by pwojcik86 — APPROVED** [review]

**Pull request merged**